cip-ledger-state.md

```diff
@@ -0,0 +1,111 @@
+ends up with twice the resource usage. The following proposal hopes to reduce resource usage and
+complexity for chain following applications like db-sync.
+
+# Current Situation
+
+Currently there is a local chain sync protocol which is really just the peer-to-peer protocol
+using a local domain socket rather than the TCP/IP socket normally used for P2P transport.
+
+The data transported over this local chain sync protocol is limited to block chain data. However, a
+Cardano node also maintains ledger state which includes:
+
+* The current UTxO state.
+* Current amount of ADA delegated to each stake pool.
+* Which stake address is currently delegated to each pool.
+* Rewards account balances for each stake address.
+* Current protocol parameters.
+
+The first of these ledger state components is by far the largest component and is probably not
+needed outside the node (and definitely not needed by db-sync. However the others are needed and
+stored by `cardano-db-sync` which gets these data sets by maintaining its own copy of ledger state
+and periodically extracting the parts required.
+
+This means that when `node` and `db-sync` are run on the same machine (which is the recommended
+configuration) that machine has two basically identical copies of ledger state. Ledger state is a
+*HUGE* data structure and the mainnet version currently consumes over 10 Gigabytes of memory.
+Furthermore, maintaining ledger state duplicates code functionality that is in `ouroboros-consensus`
+and the maintenance of ledger state has been the cause of about 90% of the bugs in `db-sync` over
+the last two years. The maintenance of ledger state also makes updating, running and maintaining
+`db-sync` by operators more difficult than it should be. Finally, if `db-sync` did not have to
+maintain ledger state, the size of the `db-sync` code base would probably decrease by about 50% and
+the bits removed are some of the most complicated parts.
+
+
+## Specification
+
+The proposed solution is an enhanced local chain sync protocol that would only be served over a
+local domain socket. The enhanced chain chain sync protocol would include the existing block chain
+data as well as events containing the ledger data that db-sync needs. This enhanced local chain
+sync protocol is useful for many applications other than just db-sync.
+
+Smart contract developers would like an application that turns block chain and ledger state changes
+into an event stream. With this enhanced local chain sync protocol, generating an easily consumable
+event stream simply requires a conversion of the binary enhanced local chain sync protocol into
+JSON.
+
+This enhanced local chain sync protocol is basically the data that would be provided by the
+proposed Scientia program (which as far as I am aware has been dropped).
+
+The ledger state data that would provided over the extended chain sync protocol is limited to:
+
+* Per epoch stake distribution (tickle fed in a deterministic manner).
```

Comment:
I think yes. It is very useful to request only 1 of mark, set, go. @papacarp at pooltool needs `mark`. At dripdropz, I need just the `go` snapshot.

Reply:
db-sync extracts slices of ~2000 per block from the `set` snashot starting from the epoch boundary. The size may be bigger based on the expected number of blocks in the epoch.

https://github.com/input-output-hk/cardano-db-sync/blob/master/cardano-db-sync/src/Cardano/DbSync/Era/Shelley/Generic/StakeDist.hs#L44

Reply:
For the sake of simpilicity, I really think this should be push only from the `node` to the consumer..

The live up-to-date distribution is not currently needed by `db-sync`.

As for which one (and making sure I have this correct, `mark` is the stake distributions that will be used in `current_epoch + 2`, `set` in `current_epoch + 1` and `set` in the `current_epoch`). My understanding is that `set` is not guaranteed stable in the first part of an epoch, but I would suggest this is the one we want, but only as soon as it is stable.

Reply:
> As for which one (and making sure I have this correct, `mark` is the stake distributions that will be used in `current_epoch + 2`, `set` in `current_epoch + 1` and `set` in the `current_epoch`). My understanding is that `set` is not guaranteed stable in the first part of an epoch, but I would suggest this is the one we want, but only as soon as it is stable.

Slight correction here:
mark = current_epoch + 1 (this is next epochs stake distribution)
set = current_epoch (this is the stake distribution being used CURRENTLY to make blocks)
go = current_epoch - 1 (this is used for reward calculations)

mark snapshot would not be guaranteed until after the stability window I presume but it would be good to pull whenever  the consumer wants as each application will have different accuracy requirements.

Reply:
Sorry, about half the time I look at the `mark|set|go` terminology I get it wrong.